### PR TITLE
general code improvement

### PR DESCRIPTION
**Pull request by drunkirishcoder**
Ok this is a general improvement PR. A few things in this one.

1. Finally untangled the `MempoolMap`. The way it was done was really hard to understand. Now instead of keeping a hashmap of mempools in the runtime when that's not being used at all, it's just a vec. Only the builders need to borrow the vec as a hashmap temporarily.

2. Fixed the order of the EAL cleanup and make ports and mempools drop before the main runtime.

3. No longer exposing the `Mempool` struct outside the crate even for testing. Instead used a testils wrapper to hide the underlying DPDK structs. They expose too much internal functionalities and better to not be leaked.

### CODE REVIEW

File: core/src/runtime/mod.rs [hunk 1]
```diff
@@ -509,6 +522,17 @@ impl Runtime {
 
 impl Drop for Runtime {
     fn drop(&mut self) {
+        // the default rust drop order is self before fields, which is the wrong
+        // order for what EAL needs. so we use a trick here to take ownership of
+        // the inner fields and make them drop first.
+        self.ports.take();
```
Comment (zeeshanlakhani) [inline core/src/runtime/mod.rs]: brought up the idea of using [manuallydrop](https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html) here instead of the options + take approach.

File: core/src/testils/mod.rs [hunk 2]
```diff
@@ -13,16 +13,36 @@ pub mod byte_arrays {
...
+}
+
+impl Drop for MempoolGuard {
+    fn drop(&mut self) {
+        MEMPOOL.with(|tls| tls.replace(ptr::null_mut()));
+    }
+}
+
+/// Creates a new mempool for test.
+pub fn new_test_mempool(capacity: usize, cache_size: usize) -> MempoolGuard {
```
Comment (zeeshanlakhani) [inline core/src/testils/mod.rs]: minor: I'm ok w/ this, but if we stick w/ this approach, I'd say name it `new_mempool` as it is exported through testils already. Otherwise, would it be cleaner to have a compiled-under-test `new` static function under mempool (w/ the guard) that gets used, and then it's just the ame new but for tests? I'm not a stickler for it though and know how much you love feature flags :).

File: core/src/testils/mod.rs [hunk 3]
```diff
@@ -13,16 +13,36 @@ pub mod byte_arrays {
...
 pub fn cargo_test_init() {
     TEST_INIT.call_once(|| {
-        eal_init(vec!["nb2_test".to_owned()]).unwrap();
+        dpdk::eal_init(vec!["nb2_test".to_owned()]).unwrap();
     });
 }
+
+/// A handle that keeps the mempool in scope for the duration of the test. It
+/// will unset the thread bound mempool on drop.
+#[allow(dead_code)]
```
Comment (zeeshanlakhani) [inline core/src/testils/mod.rs]: why is this here if it's used within? does the macro not pick it up then?

**Comment by drunkirishcoder** [comment]
yeah the lib needs more coverage. but too much of the code depend on ffi and not sure how best to test them. not the ideal way of doing things but we will need to revisit and beef that up for sure.

File: core/src/testils/mod.rs (hunk 3 above)
Comment (drunkirishcoder) [reply core/src/testils/mod.rs]: the field `inner` is never used by anything, but just held onto so it lives for the duration of the test.
